extension/source/background/KeyringController.ts

```diff
@@ -0,0 +1,111 @@
+import { BlsWalletWrapper } from 'bls-wallet-clients';
+import { ethers } from 'ethers';
+import generateRandomHex from '../helpers/generateRandomHex';
+import { NETWORK_CONFIG } from '../env';
+import QuillStorageCells from '../QuillStorageCells';
+import assert from '../helpers/assert';
+import { PartialRpcImpl, RpcClient } from '../types/Rpc';
+import ensureType from '../helpers/ensureType';
+
+export default class KeyringController {
+  constructor(
+    public InternalRpc: () => RpcClient,
+    public keyring: QuillStorageCells['keyring'],
+    public selectedAddress: QuillStorageCells['selectedAddress'],
+    public ethersProvider: ethers.providers.Provider,
+  ) {}
+
+  rpc = ensureType<PartialRpcImpl>()({
+    eth_accounts: async ({ origin }) => {
+      if (origin === window.location.origin) {
+        return (await this.keyring.read()).wallets.map(
+          ({ address }) => address,
+        );
+      }
+
+      const selectedAddress = await this.selectedAddress.read();
+
+      // TODO (merge-ok) Expose no accounts if this origin has not been
+      // approved, preventing account-requiring RPC methods from completing
+      // successfully only show address if account is unlocked
+      // https://github.com/web3well/bls-wallet/issues/224
+      return selectedAddress ? [selectedAddress] : [];
+    },
+
+    eth_requestAccounts: async (_request) => {
+      const selectedAddress = await this.selectedAddress.read();
+      const accounts = selectedAddress ? [selectedAddress] : [];
+      return accounts;
+    },
+
+    eth_coinbase: async (_request) =>
+      (await this.selectedAddress.read()) ?? null,
```

Comment:
I'm pretty sure this should get the current block’s beneficiary address.

```suggestion
    eth_coinbase: async (_request) => {
      const blockNum = await this.ethersProvider.getBlockNumber();
      const block = await this.ethersProvider.getBlock(blockNum);
      return block.miner;
    }
```

Reply:
Issue created